BREAKING CHANGE: 2.0 - Preview

cecilia-sanare opened the pull request:
**Changes**

- Switched from callbacks to promises
- Requests are now intercepted directly
  through XHR and Fetch
- Added a "provider" concept to support multiple
  authentication gateways (auth0, cognito, wso2, etc.)
- Switched from PhantomJS to Chrome Headless
- Updated Libraries and switch off of deprecated libraries
- Added support for custom query params
- Added Compat to track browser api support

This closes #21, closes #94, closes #95, closes #41, closes #5

Commit:
BREAKING CHANGE: 2.0 - Preview

- Switched from callbacks to promises
- Requests are now intercepted directly
  through XHR and Fetch
- Added a "provider" concept to support multiple
  authentication gateways (auth0, wso2, etc.)
- Switched from PhantomJS to Chrome Headless
- Updated Libraries and switch off of deprecated libraries
- Added support for custom query params
- Added Compat to track browser api support

davewoodward commented on src/providers/cognito.js:
```diff
@@ -0,0 +1,11 @@
+class SalteAuthCognitoProvider {
+  static deauthorizeUrl(config) {
+    const gateway = config.gateway.replace('/oauth2', '');
```
I think we should only accept the base URL as a configuration parameter and then assume that things like "oauth2" have to be appended at the specific provider level.  I realize that this will require both the AuthO and WSO2 providers to append "oauth2" but I still think that would be better than having providers like Cognito have to remove them.

cecilia-sanare replied:
Every other provider already expects the `gateway` to be the base url.

Cognito is the exception since its `/authorize` endpoint is hidden under a `/oauth2` path.

As a compromise, I think we should allow `providers` to override the `authorizeUrl` function.

If a provider doesn't implement it, then we'll just append `/authorize` to the `gateway` url.

davewoodward commented on src/salte-auth.profile.js:
```diff
@@ -0,0 +1,237 @@
+      for (let i = 0; i < params.length; i++) {
+        const param = params[i];
+        const [key, value] = param.split('=');
+        this.parse(key, value);
```
In the pop-up scenario, the parameters saved to session storage will not be available to the parent window.

Commit:
chore: various fixed

- Support switching the storageType
- Support transfering storage for popups

Commit:
chore: add support for overriding authorizeUrl

Commit:
docs: added documentation

Commit:
chore: added more validation

- Added validation for audience and azp
- Cleaned up tokens in the tests to be easier to read
- Added documentation for "storageType"

Commit:
docs: update example on readme

Commit:
docs: added documentation url

cecilia-sanare commented on src/salte-auth.js:
```diff
@@ -1,1226 +1,308 @@
-import uuid from 'uuid';
-
-export default class AuthenticationContext {
+/** Salte Auth */
```
What should the documentation for the `SalteAuth` class say?

cecilia-sanare commented on package.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "salte-auth",
+  "name": "@salte-io/salte-auth",
```
Should we call it `@salte-io/salte-auth` or `@salte-io/auth`?

cecilia-sanare replied:
Decided to keep it `@salte-io/salte-auth`

cecilia-sanare commented on package.json:
```diff
@@ -24,50 +24,57 @@
     "build:dev": "webpack --progress --colors -d",
     "build:prod": "webpack --progress --colors -p",
     "report-coverage": "coveralls < coverage/lcov.info",
-    "publish-latest": "publish-latest --user-email bot@salte.io --user-name salte-bot",
```
Should we continue to support installs via bower?

cecilia-sanare commented on src/providers/cognito.js:
```diff
@@ -0,0 +1,25 @@
+/** Provider for Amazon's Cognito */
+class SalteAuthCognitoProvider {
```
We need to verify that cognito works.

cecilia-sanare replied:
Cognito has been verified.

cecilia-sanare commented on src/salte-auth.profile.js:
```diff
@@ -0,0 +1,353 @@
+        code: 'invalid_aud',
+        description: 'The audience did not match the Client ID.'
+      };
+    }
```
Should we verify that the other claims are present? (username, email, etc)

Commit:
fix: external dependencies

Commit:
chore: add the ability to disable validation

- added support for provider defaults

davewoodward commented on src/salte-auth.js:
```diff
@@ -1,1226 +1,323 @@
+    window.salte.auth = this;
+
+    if (!config) {
+      throw new ReferenceError('A config must be provided.');
```
All configuration errors should be identified before we set window.salte.auth to this.  Otherwise, subsequent calls may result in us returning a partially configured component.

Commit:
chore: test scenario where config isn't set when initializing salte-auth

Commit:
docs: updated esdocs

Commit:
chore: renamed signIn/signOut to login/logout

- 100% documentation coverage is now required

Commit:
fix: full page reload on redirect

davewoodward commented on src/salte-auth.js:
```diff
@@ -1,1226 +1,324 @@
-    frameHandle.src = 'about:blank';
-    this._loadFrameTimeout(urlNavigate, 'authRenewFrame' + resource, resource);
-  }
+  get authorizeUrl() {
```
I think this should be renamed to idTokenUrl to be consistent with the getter for accessTokenUrl.  It is also confusing because both salte-auth and the provider can currently have something called an authorizeUrl.

Commit:
feat: added support for azure ad

- url parameters are now decoded

Commit:
chore: renamed various functions

- authorizeUrl -> loginUrl
- authorizeUrl -> authorizeEndpoint

Commit:
fix: added redirectErrorCallback

Commit:
chore: privatized a bucnh of methods / properties

Commit:
docs: breaking changes documentation

Commit:
chore: bundle uuid with salte-auth

Commit:
chore: greenkeeper

Commit:
docs: bower

Commit:
chore: renamed redirectErrorCallback to loginCallback

Commit:
chore: public access for scoped package

davewoodward commented on src/salte-auth.js:
```diff
@@ -1,1226 +1,358 @@
- *  @property {object} profile - properties parsed from idtoken.
+ * The configuration for salte auth
+ * @typedef {Object} Config
+ * @property {String} gateway The base url of your identity provider.
```
We should probably change this from gateway to identityProviderBaseUrl or something similar.  It is just a happenstance that WSO2 serves as both an identity provider and an API  gateway.

davewoodward commented on src/salte-auth.js:
```diff
@@ -1,1226 +1,358 @@
+ * @property {String} redirectUrl The redirect url specified in your identity provider.
+ * @property {String} clientId The client id of your identity provider
+ * @property {String} scope A list of space-delimited claims used to determine what user information is provided and what access is given. Most providers require 'openid'.
+ * @property {Boolean|Array<String>} routes A list of secured routes. If true is provided then all routes are secured.
```
Do you think that routes makes it clear that this is referring to application navigation routes whereas endpoints refers to API endpoints?

cecilia-sanare replied:
The term route is fairly common when referring to an applications routes.

**References**

- https://www.webcomponents.org/element/PolymerElements/app-route
- https://docs.angularjs.org/api/ngRoute/provider/$routeProvider
- https://github.com/ReactTraining/react-router
- https://router.vuejs.org/en/

Commit:
chore: added failing tests for next commit

Commit:
chore: fix tests

cecilia-sanare commented:
**A few issues;**

- [ ] We need to add support for all input values for fetch #98
- [ ] We need to figure out why our tests have issues running multiple times (e.g. `yarn run tdd`) #100
- [ ] We need to completely isolate our tests from one another #99

**References**

https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch

Commit:
docs: demos

- renamed gateway to providerUrl
- added cognito demo

Commit:
chore: removed momentjs

cecilia-sanare commented:
This PR should be good to go, we just need to make sure we give it a meaningful squash message.

Commit:
chore: added failing tests for idToken exp

Commit:
chore: fixed failing tests for idToken Exp

Commit:
chore: added commonjs2 to webpack

Commit:
fix: routes end endpoints should use relative urls
